### PR TITLE
#163388254 Add CRUD functionality for articles

### PR DESCRIPTION
**Pull request by calebrotich**
#### What does this PR do?
Allows a user to perform CRUD actions on articles using their relevant permissions

#### Description of Task to be completed?
- add tests for articles CRUD
- add ability of user to create, read, update and delete articles
- control user permissions on articles

#### How should this be manually tested?
- Make the following requests to _postman_:

| Request         | URL                | Description |
| ------------- | ------------- |------------------|
| `POST`  | `{BASE_URL}/api/articles/ ` |Create a new article|
| `GET`  | `{BASE_URL}/api/articles/`  |Fetch all articles |
| `GET`  | `{BASE_URL}/api/articles/<slug> ` |Fetch a specific article using its slug|
| `PUT`  | `{BASE_URL}/api/articles/<slug> ` |Update details of a specific article|
| `DELETE`  | `{BASE_URL}/api/articles/<slug> ` |Delete an article|

#### What are the relevant pivotal tracker stories?
[#163388254](https://www.pivotaltracker.com/story/show/163388254)

#### Screenshots (if appropriate)
_**Post an article**_
image

_**Get All articles**_
image

_**Get Specific Article**_
image

_**Edit Article**_
image

_**Delete Article**_
image

### CODE REVIEW

**Comment by TooColline** [comment]
It works fine as expected @calebrotich10. Maybe just make the dates format user-friendly.

**Comment by unah254** [comment]
Nice work! I believe the author's id will be changed to the author's name once my PR is approved

**Review by unah254 — COMMENTED** [review]
Works well as expected! However, I feel you should use an author's name instead of id on the response. This will be more appropriate than just seeing an id 1

**Review by bryan-munene — APPROVED** [review]
Works as expected. Just rebase to have one commit and we are good to go

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class Article(models.Model):
+    """
+    The model for creating the articles is defined here
+    """
+    slug = models.SlugField(
+        db_index=True, max_length=1000, unique=True, blank=True)
+    title = models.CharField(max_length=1000, blank=False)
+    description = models.CharField(max_length=2000, blank=False)
+    body = models.TextField(blank=False)
+    image_url = CloudinaryField(blank=True, null=True)
```
Comment (sam-karis) [inline authors/apps/articles/models.py:17]: Ensure you save both the image and the absolute image URL after the image is uploaded to cloudinary.

**Review by sam-karis — CHANGES_REQUESTED** [review]

File: authors/apps/articles/tests/base_test.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        response = self.client.post(self.login_url,
+                                    self.test_user,
+                                    format='json')
+
+        token = response.data['token']
+        return token
+
+    def create_article(self):
+        token = self.login_user()
+        url = self.create_list_article_url
```
Comment (sam-karis) [inline authors/apps/articles/tests/base_test.py]: I don't the reassignment of `self.create_list_article_url` to `url` is necessary just use `self.create_list_article_url` directly.

File: authors/apps/articles/tests/test_articles.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        self.base_articles(message, response)
+
+    def test_article_update(self):
+        """
+        This method checks if a user can update an existing articles
+        """
+        response = self.authorized_post_request(self.create_list_article_url)
+        slug = response.data['slug']
+        token = self.login_user()
+        data = {
```
Comment (sam-karis) [inline authors/apps/articles/tests/test_articles.py]: I recommend you put the data in the base test to keep your test clean and concise.

File: authors/apps/articles/tests/test_articles.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        self.http_200_ok(res)
+        message = "Test description for the article"
+        self.base_articles(message, response)
+
+    def test_unauthorised_user_update(self):
+        """
+        This method tests if unauthorized user can update existing articles
+        """
+        slug = self.create_article()
+        data = {"article": {"title": "Updated Title", "body": "Updated body"}}
```
Comment (sam-karis) [inline authors/apps/articles/tests/test_articles.py]: Same as above

**Review by sam-karis — CHANGES_REQUESTED** [review]

**Review by sam-karis — APPROVED** [review]
LGTM

**Pull request merged**